### DISCUSSION TITLE
Less boilerplate code with the new primary constructor in C# 12

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#99] [General]
# Less boilerplate code with the new primary constructor in C# 12

The new language feature "primary constructor" which will be released with C# 12 / .NET 8 this year (November 2023) allows you to remove some ceremonial code. Let's see how.

https://steven-giesel.com/blogPost/1b2a4f18-86da-42d3-9ddc-8b41ed1eba0f/less-boilerplate-code-with-the-new-primary-constructor-in-c-12

**Comment by toupswork** [comment]
I use this all the time. I love how succinct it is. I just think they really dropped the ball by not allowing the ability to mark the parameters as readonly. I'm sure they will in a future version.

**Reply by linkdotnet** [reply]
I hope so too.
In my current project we explicitly ruled against them - not because of the readonly flag, but because we have some interns snd it might be overwhelming.
Personally I wished they would have gone more the TypeScript way of primary constructors where you can directly declare them readonly and or private/protected/…

**Reply by toupswork** [reply]
exactly. I have guys on my team who were hired on after interning a year ago, and they don't have any confusion. But they were also used to using Records

**Reply by linkdotnet** [reply]
Do you have a rule when not to use them?
Like if you have more than 3 dependencies or so?

to be honest while I get the „problem“ with the missing readonly, I never had an issue that a dev reassigned a service injected by the container
That, of course, doesn’t invalidate the point, but from m point of view there are other issues to tackle.
Did you ever come across that issue?

**Reply by toupswork** [reply]
No I don't have rules like that. I am all for succinctness. It's tedious to write constructors all day and assign the parameter to readonly fields. And then refactoring and having to change them or adding more dependencies. That's all just noise that gets in the way of focusing on the problem I am trying to solve. No, I never had an issue with a variable being re-assigned, but that's not the thing about readonly that I am missing. There's a performance boost to having them marked as readonly.

**Reply by linkdotnet** [reply]
When did they introduce that?

I remember in IL Code there was no sign of readonly anymore (with net5 or net6 IIRC).
That said using readonly or not leads to the same code - hinting towards no performance benefit. 
I even have a blog post about it (will link it afterwards).

And even  ow you can reassign stuff via reflection.

**Reply by linkdotnet** [reply]
Oh I never did a blog post … Thought I did.
My point still stands.
Currently only on mobile (on vacation) so I have to check once I am back in a few days. 

I am almost certain that the IL code was the same in older versions.

**Reply by toupswork** [reply]
You know I could have sworn I read that it increases performance. I have known for a while that it's really not readonly at the IL Level and you can simple use reflection to change it. But I know I read it was faster. I am going to try to find an article on that right now.. If anything, the JIT could optimize it to be pulled in on the cache line for multiple cores because there is no worry that it will change.

**Reply by linkdotnet** [reply]
I quickly checked (a maybe oversimplified) example and it adds „just“ the initonly flag (since C# introduced init).
Have to play around with the JIT Code.

But yeah if the IL or JIT Code is the same I wouldn’t know where the benefit should come from!?

let me know if you she’s some findings.

**Reply by toupswork** [reply]
Well I looked, and I can't find anything official. But I do see the following differences in IL below. I couldn't find anything where someone tried to test the performance.. That could be a cool thing to try out with Benchmark.  But just think about it like this, if JIT sees that it can't be changed, then it's not going to bother worrying about volatile reading. It might ensure it gets cachelined. 
```
  .field private initonly object readOnlyObj

  .field private object notReadOnly
```

**Reply by linkdotnet** [reply]
The only sources I find indicate just compiler behaviour.

I will check this later once I have access to my laptop
